### ISSUE TITLE
Load previously open database on startup doesn't work

### ISSUE THREAD

**Issue by Major-Mike** [keepassxreboot/keepassxc#8745]
## Overview
[TIP]:  # ( DO NOT include screenshots of your actual database! )
[NOTE]: # ( Give a BRIEF summary about your problem )
Since the last update to version 2.7.4 the open databases that are "checked" to be remembered are not loaded automatically when you reboot the system and KPX loads.

## Steps to Reproduce
[NOTE]: # ( Provide a simple set of steps to reproduce this bug. )
1. Go to Tools > Settings
2. Select General > Basic Settings
3. Check "Remember previously used databases"
4. Check "Load previously open databases on startup"
5. Check "Remember database key files and security dongles" (optional?)
6. Click OK and open up a database, unlock it and use it
7. Shutdown/Reboot the system and start up again
8. Previous open database doesn't load

## Expected Behavior
[NOTE]: # ( Tell us what you expected to happen )
Like before, the databases that were last used would load up the next time.

## Actual Behavior
[NOTE]: # ( Tell us what actually happens )
Previously used databases are not loaded next time.

## Context
[NOTE]: # ( Give us any additional information you may have. )

[NOTE]: # ( Paste debug info from Help → About here )
KeePassXC - Version 2.7.4
Revision: 63b2394

Qt 5.15.6
Debugging mode is disabled.

Operating system: Windows 10 Version 2009
CPU architecture: x86_64
Kernel: winnt 10.0.22621

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare
- YubiKey
- Quick Unlock

Cryptographic libraries:
- Botan 2.19.1

[NOTE]: # ( Pick choices based on your environment )
Operating System: Windows
Desktop Env: 10/11
Windowing System: N/A

**Comment by droidmonkey** [comment]
It works fine. Make sure you aren't closing the database prior to shutting down or closing keepassxc. Also try deleting the keepassxc.ini file from c:/users/yourusername/appdata/local/keepassxc

**Comment by Major-Mike** [comment]
No, I am not closing them, I am doing the same thing I did before the update, so my behavior hasn't changed and it was working fine before. I don't think it is appropriate to summarily close an issue without some degree of effort debugging, might work for you or some but that doesn't mean it is working for everyone.

**Comment by droidmonkey** [comment]
Is your database on a network share?

This feature is enabled by default and a fundamental feature. If it was systemically broken we would be getting hundreds of reports. It is not broken and the code for it hasn't changed for years. So it's something with your specific setup and we don't keep issues opened for user specific faults. If we do find it's a code fault then I'll reopen. That's the only way we can keep our issue list contained.

**Comment by Major-Mike** [comment]
No, it is not. 

And I hear you, but when nothing about a setup changes except that it was just updated, what else is one to assume? You tell me and I will assume that.

And as a developer, I can tell you that features can break without being updated themselves by something else, just something to consider. I also never subscribed to the, if it was broken would have hundreds complaining, some may not care, some may not want to, and some may not notice. For days I reopened them thinking maybe I did something until I realized it consistently doesn't open up. 

TL/DR
Your audience is not as broad as some of the mainstream managers, so perhaps the outcry might match the number of people using it. Anyway, if you ask and I can provide something to help, let me know and if you are content saying there is no problem and ignoring it until hundreds of people report it which may never come, then I guess nothing I can do.

**Comment by michaelk83** [comment]
If you're a developer, your best bet would be to try to debug it yourself and submit a PR. Even if it's broken on their end, they can't debug it if they can't reproduce it.

**Comment by droidmonkey** [comment]
Please provide the following information:
* Is your recent database list populated with files?
* Check your appdata/local/keepassxc/keepassxc.ini file and see if the `LastOpenedDatabases` entry exists and has a value

For reference the code that opens previous databases is from 2 years ago. 

When MainWindow is created, at the end it restores the previous configuration state: https://github.com/keepassxreboot/keepassxc/blob/3cbe4df8c753bc17197cf8615948a275ea524d96/src/gui/MainWindow.cpp#L690

Which then uses the database files in the LastOpenedDatabases config variable to call standard openDatabase function https://github.com/keepassxreboot/keepassxc/blob/3cbe4df8c753bc17197cf8615948a275ea524d96/src/gui/MainWindow.cpp#L713

**Comment by Major-Mike** [comment]
> If you're a developer, your best bet would be to try to debug it yourself and submit a PR. Even if it's broken on their end, they can't debug it if they can't reproduce it.

@michaelk83 Thank you for the input and I am almost certain you don't know this but just because someone is a developer, it doesn't mean they can come into any project and instantly know what is what and start making changes. That is an absurdly offensive assumption in the face of the people who spend their time managing the code to say that someone who has not worked on this code and would need substantial time to get up and familiar with it can just roll in and make a change just like that like they know better. The process is always served better to provide them details, logs, whatever you can muster so that they being familiar with their code might get inspiration where to look.

@droidmonkey Please forgive the delay getting back to you, I was out of town for Veteran's day visiting families. Answer to your first question, yes, it shows the files - refer below:

![image](https://user-images.githubusercontent.com/55165479/201841101-11811768-4b37-4175-8767-107180c68e69.png)

In answer to your second question, this is what it shows (I have pruned some content):

```
[General]
LastOpenedDatabases=@Invalid()
LastDir=@Variant(\0\0\0\...0y\0n\0\x63)
LastDatabases=C:\\Users\\xxxk.kdbx, C:\\Users\\xxxd.kdbx, C:\\Users\\xxxl.kdbx
LastKeyFiles=@Variant(\0\0\0\...x65\0y\0x)
LastChallengeResponse=@Variant(\0\0\0\...0\0\0)
```
I am not sure what the context of `@Invalid()` is but the path of those 3 files are valid, they do exist. I assume trying to `.toStringList()` on this would result in null and therefore no filenames for it to open up, right?

I understand, I believe you that it hasn't been updated, just for me the only change was the update, so it seemed like a valid assumption but could have been coincidental in timing. I will share one thing that I have observed that might help, not sure.

The kdbxs are shared between a desktop install and a portable instance. The files are synced between the two and so when I open them using the portable at work, I have the latest version and when I open it at my home desktop, it is also up to date. Since I was out of town, I only used the portable and the first time I ran the application, they did not open up, did it manually. But after that each time they did open up automatically. Could it be that if we switch between instances this feature breaks? To test this, when I got home, I tried to use the desktop version, they did not open up. I opened them and then the next time, they all opened up. But when I ran at work with my portable, they did not open up, when I got home, the desktop version didn't open up. So it feels like using different instances with the same files behaves this way. 

Thought that might hopefully be a helpful observation. Let me know if I can provide anything else. Thank you again for taking the time. I do appreciate the effort and if I didn't care about the project, I wouldn't persist trying to figure it out. I hope you will bear with it. Take care.

**Comment by michaelk83** [comment]
> Thank you for the input ...

Coming in from outside and proposing new code contributions is the whole point of open source. Yes, it takes some time to get familiar with a new code base, but you don't need to learn all of it to make a useful contribution. The review process is there for a reason: if the maintainers don't like your changes, they will either request changes, or reject your PR. You are of course welcome to consult the maintainers if you are unsure how to proceed with a proposed change.

I am a developer myself, and I would not be offended at all if someone were to propose _useful_ changes to my code, especially if I've open sourced it. On the contrary, I've been following this and other open source projects for a while now, and I can tell you for a fact that PRs are almost always [welcomed](https://github.com/keepassxreboot/keepassxc/blob/develop/.github/CONTRIBUTING.md), and more so than another issue to debug or develop. Maintainers have limited time, and too much to do as it is.

(Sorry for taking this off-topic.)

**Comment by droidmonkey** [comment]
Does your portable instance have a config directory and a `.portable` file? You may need to just re-download the portable version and unzip it on to your drive. It sounds to me like it is writing config files to your system and not the portable directory.

**Comment by Major-Mike** [comment]
@droidmonkey Jonathan sorry for the late reply, it was a rough year's end and I had to travel for work, but finally did something that fixed it not sure why but figured I share it back in case it helps you any.

Logged into KeePassXC, DB loaded, open and unlocked, if I shut down the machine, it would not show up again which is nuts. But on a whim decided one day to "intentionally" exit the whole thing and THEN shut down the machine. Next time I loaded the machine, it was there, it had loaded it, opened it and I only had to type in the password for the first time before the biometrics could be used and now since I did that, I can just shut down the machine locked or unlocked without anything special and it will come back, so not sure why it was `@Invalid()` but now loads just fine, I have no idea why

I can only guess that between the system shutting down sending "term" signal and it "writing" the last path to DB, it was losing the race condition, hence `@Invalid()` but since the intentional exit allowed it to finish writing the path, now even though the race condition may still exist and it can't "update it" it won't change what is already there either, and being and actual path, opens up. not sure if that's what happened but if I had to make a guess.

**Comment by jstiete** [comment]
> to "intentionally" exit the whole thing and THEN shut down the machine.

That also seems to work for me. Thanks!

**Comment by Major-Mike** [comment]
I am glad it helped you.

**Comment by ygoe** [comment]
So, this issue still exists but is closed. I sometimes also run into this. Maybe always but I don't always notice. I have a regular installation with nothing special, a locally stored database that no other process touches, and only a single database. It's the only entry in the list of recently used databases when KeePassXC starts. Yet it won't load this database at the automatic startup after login. The option "Load previously open databases on startup" is checked. But it clearly does not have the described effect. And please don't tell me what I see is not true.

You'll probably also request me to patch your code, but unfortunately I can't do that. Yes, I'm a developer, but not for C++ or Qt. AFAIR, Qt isn't even freely available anymore so I can't use it anyway. And last time I did, it was the main source of memory bugs and app crashes. Even before C++ itself. I'm not into that kind of fun.

**Comment by droidmonkey** [comment]
Are you even leaving the database open when you close the program. Your report makes it sound like you have closed it.

**Comment by ygoe** [comment]
I have not touched KeePassXC before shutdown. So, am I leaving it open? No, I'm not doing anything. Am I closing it? Certainly not. The question is rather, what is the application doing? I can't watch it during Windows shutdown. But I would assume that the user wouldn't have to care. The option label is clear and that's precisely what I expect from it.

**Comment by jcasale** [comment]
I've also had the problem a few times and each time it was due to a configuration file with an invalid configuration.

However, it was not the file in `AppData\Local\KeePassXC\keepassxc.ini` and instead the file in `AppData\Roaming\KeePassXC\keepassxc.ini` that I needed to remove.

**Comment by ygoe** [comment]
Alright, I have edited that file in %LocalAppData% and removed the line with the "Invalid" value. (After manually closing KeePassXC.) Then I opened it again, manually selected and opened the database and again closed the application. It was only now that two other lines were added to that file.

It seems to work now, the database is automatically loaded at startup.

From my observations I guess there's two issues here.

1. The usual wrong programming that settings files are not written after a setting has changed (in this case an opened database) but only when the application is successfully terminating on its own. Almost all applications have this bug. It leads to data loss in case of an unclean shutdown. See the next point.
2. When rebooting Windows, KeePassXC doesn't seem to have enough time to write its settings file with the currently loaded database. This bug is usually invisible because once you have opened a database and then manually closed the app, the file will be written. You simply won't notice anymore that the file will never be written again on the subsequent Windows shutdowns if you're not making any changes to the settings. Which probably happen rarely, at least for me.

**Comment by GuardianMajor** [comment]
Actually the way I have fixed it before was to intentionally close it BEFORE shutting down so it will register the last DB, then open it again, then shutdown; then next load it comes up. The issue seems to be that during shutdown there is a race condition where it hasn't always completed the write before the system terminates it. You can test this by having multiple databases and it will open an older one because it has it but not a more recent one that was closed. Anyway can be annoying but not the end of the world.

**Comment by Squaredude** [comment]
Had this happen on Pop_OS! (Linux) and @Major-Mike's solution fixed it for me. Thanks a lot!

**Comment by Major-Mike** [comment]
Glad it helped you.
